keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,1100 @@
+- [Flux Operator](https://flux-framework.org/flux-operator/)
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- allow to mark a job as a `Succeeded` even if some of the indexes fail.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Change the existing behavior of `NonIndexed` mode.
+- Change the existing behavior of `Indexed` mode without a SuccessfulPolicy.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+We propose new policies under which a job can be declared as succeeded.
+Those policies can be modeled in the following:
+
+1. An indexed job completes if a set of [x, y, z...] indexes are successful.
+2. An indexed job completes if x or x% of indexes are successful.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a machine-learning researcher, I run an indexed job which drivers (leaders) are launched
+as index=0,1,2, and workers are launched as non index=3+.
+I want to care about only the drivers when the result of job is evaluated.
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  successfulPolicy:
+    requiredIndexes: "0-2"
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: job-container
+        image: job-image
+        command: ["./sample"]
+```
+
+#### Story 2
+
+As a machine-learning researcher, I run an indexed job with an autoscale semantic
+set as a minimum number is 5 and a maximum number is 20.
+I want to mark the job as successful at least 5 indexes succeeded, even if 
+all indexes don't succeed.
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  successfulPolicy:
+    minSucceededIndexes: 5
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: job-container
+        image: job-image
+        command: ["./sample"]
+```
+
+#### Story 3
+
+As a machine-learning researcher from the [Story 2](#story-2), I want to mark the job
+as successful if 50% of all indexes succeed because I can not know how many indexes 
+will have in advance.
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  successfulPolicy:
```

Comment:
Overall, I prefer your suggestion. But `list` sounds weird to me since the form (e.g., "0-2,3,10") isn't a list.
How about `compressedList`? Anyway, discussing API design with API reviewers would be better.

```
successPolicy:
  onSucceededIndexes:
    compressedList: "0-2"
    atLeast: "5%"  # only one of the fields is allowed
```

Reply:
```yaml
successPolicy:
  onSucceededPods:
    indexes: "0-2"
    count: "5%"
```

This could work better. But again, for API review.

Note that enhancements freeze is tomorrow.

Reply:
Thanks for the suggestions and notification.
I will do my best as much as possible.